src/nvim/mark.c

```diff
@@ -221,245 +228,402 @@ void checkpcmark(void)
- */
-pos_T *getmark_buf(buf_T *buf, int c, bool changefile)
+/// Get a named mark.
+///
+/// All types of marks, even those that are not technically a mark will be returned as such, to then
+/// move to the mark use mark_move_to().
+/// @param buf  Buffer to get the mark from.
+/// @param win  Window to get or calculate the mark from (motion type marks, context mark).
+/// @param flag  kMarkAll (default) returns any mark found, if global it resolves it's fnum.
+///              kMarkBufLocal return only marks that belong to buf, kMarkAllNoResolve same as
+///              kMarkAll but does not resolve the fnum.
+///
+/// @return          Mark, if not found returns NULL, if kMarkBufLocal returns NULL when
+///                  mark->fnum != buf->handle.
+/// @return[static]  Marks that need "wrapping", are statically allocated see get_local_mark,
+///                  get_motion_mark for an explanation.
+fmark_T *mark_get(buf_T *buf, win_T *win, MarkGet flag, int name)
 {
-  return getmark_buf_fnum(buf, c, changefile, NULL);
+  fmark_T *fm = NULL;
+  if (ASCII_ISUPPER(name) || ascii_isdigit(name)) {
+    // Global marks
+    xfmark_T *xfm = mark_get_global(!(flag & kMarkAllNoResolve), name);
+    fm = &xfm->fmark;
+    // Only wanted marks belonging to the buffer
+    if (flag & kMarkBufLocal && xfm->fmark.fnum != buf->handle) {
+      return NULL;
+    }
+  } else if (name > 0 && name < NMARK_LOCAL_MAX) {
+    // Local Marks
+    fm = mark_get_local(buf, win, name);
+  }
+  return fm;
 }
 
-pos_T *getmark(int c, bool changefile)
+/// Get a global mark {A-Z0-9}.
+///
+/// @note  input must be validated previously otherwise undefined behavior.
+/// @param name  the name of the mark.
+/// @param resolve  If true try resolving the mark fnum, meaning load the buffer stored in
+///                 the mark fname and update the xfmark_T (expensive).
+///
+/// @return  Mark, undefined behavior if input is not validated.
+xfmark_T *mark_get_global(bool resolve, int name)
 {
-  return getmark_buf_fnum(curbuf, c, changefile, NULL);
+  xfmark_T *mark;
+
+  assert(ASCII_ISUPPER(name) || ascii_isdigit(name));
+  if (ascii_isdigit(name)) {
+    name = name - '0' + NMARKS;
+  } else {
+    name -= 'A';
+  }
+  mark = &namedfm[name];
+
+  if (resolve && mark->fmark.fnum == 0) {
+    // Resolve filename to fnum (SHADA marks)
+    fname2fnum(mark);
+  }
+  return mark;
 }
 
-pos_T *getmark_buf_fnum(buf_T *buf, int c, bool changefile, int *fnum)
+
+/// Get a local mark (lowercase and symbols).
+///
+/// Some marks are not actually marks, but positions that are never adjusted or motions presented as
+/// marks. It searches first for marks and falls back to finding motion type marks. If it's known
+/// ahead of time that the marks is actually a motion use the function mark_get_motion directly.
+///
+/// @note  Lowercase, last_cursor '"', last insert '^', last change '.' are not statically
+/// allocated.
```

Comment:
> The freeing of memory related to marks is handled by the buffer creation, deletion logic.

The comment in the code is confusing to me because it makes me think of how I should be managing the memory that gets returned.  Your response makes it sound like the issue is more about whether the values can change out from under me, not about memory lifetimes.

If that's the case, I'd suggest rewording to focus more on this:

> The relevance is if you plan to get the value of multiple marks and you're not sure what type they are one should copy them (this is rare).

Reply:
Tweaked the documentation a bit, perhaps now is more clear.